controllers/upgrade_handlers.go

```diff
@@ -54,9 +59,86 @@ func (b BackupValidationError) Error() string {
 
 // handleUpgrade orchestrate main upgrade steps and update status as needed
 func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
-
 	r.Log.Info("Starting handleUpgrade")
 
+	currentStaterootName, _ := r.RPMOstreeClient.GetCurrentStaterootName()
+	r.Log.Info("stateroots", "current stateroot:", currentStaterootName, "desired stateroot", getDesiredStaterootName(ibu))
+
+	// Checking the stateroot to figure if pre or post reboot. WARNING: the pod may not know if we are boot loop (for now)
+	if getDesiredStaterootName(ibu) != currentStaterootName {
+		r.Log.Info("Starting pre pivot steps and will pivot to new stateroot with a reboot")
+		ctrlResult, err := r.prePivot(ctx, ibu)
+		if err != nil {
+			//todo: abort handler? e.g delete desired stateroot
+			return doNotRequeue(), nil
+		}
+
+		// prePivot requested a requeue
+		if ctrlResult != doNotRequeue() {
+			return ctrlResult, nil
+		}
+
+		err = r.rebootToNewStateRoot()
+		if err != nil {
+			//todo: abort handler? e.g delete desired stateroot
+			r.Log.Error(err, "")
+			upgradeFailedStatus(ibu, err.Error())
+			return doNotRequeue(), nil
+		}
+	} else {
+		r.Log.Info("Pivot successful, starting post pivot steps")
+		// TODO: Post-pivot steps
+		/*
+				1. oadp restore
+				    - platform
+				    - acm recovery
+
+			   2. ApplyExtraManifestsFromDir
+
+			   3. health check? readiness check /readyz endpoint?
+					- mcp
+					- clusteroperator
+					- all csv
+
+				4. restore application manifests
+		*/
+
+		r.Log.Info("Done handleUpgrade")
+
+		utils.SetStatusCondition(&ibu.Status.Conditions,
+			utils.GetCompletedConditionType(lcav1alpha1.Stages.Upgrade),
+			utils.ConditionReasons.Completed,
+			metav1.ConditionTrue,
+			"Upgrade completed",
+			ibu.Generation)
+		utils.SetStatusCondition(&ibu.Status.Conditions,
+			utils.GetInProgressConditionType(lcav1alpha1.Stages.Upgrade),
+			utils.ConditionReasons.Completed,
+			metav1.ConditionFalse,
+			"Upgrade completed",
+			ibu.Generation)
+		return doNotRequeue(), nil
+	}
+
+	return doNotRequeue(), fmt.Errorf("something we went wrong. upgrade failed")
+}
+
+// rebootToNewStateRoot reboot
+func (r *ImageBasedUpgradeReconciler) rebootToNewStateRoot() error {
+	r.Log.Info("rebooting to a new stateroot")
```

Comment:
can we also set a condition.reason or message showing the reboot has been triggered.

Reply:
It's for recording keeping by the user? Do you suggest adding a "sub" in-progress type to make it stay? Otherwise once we are done with post-pivot that timestamp is overwritten yeah?

Reply:
okay, makes sense.

Reply:
We can make it an event instead

Reply:
Added an event to indicate reboot attempt. Testing it e2e before resolving brb

Reply:
Added the event but it seems the shut down happens pretty fast. i see the event log 
```
2023-12-04T17:53:11Z	DEBUG	events	System wil now reboot	{"type": "Normal", "object": {"kind":"ImageBasedUpgrade","namespace":"openshift-lifecycle-agent","name":"upgrade","uid":"97323e1a-05df-49c2-88a4-ebfb15fe70f3","apiVersion":"lca.openshift.io/v1alpha1","resourceVersion":"31260"}, "reason": "Reboot"}
```
but client-go cant see to post it fast enough (cant see it when i watch describe)